File: backend/src/nyaa_scraper.py
from bs4 import BeautifulSoup
import requests


class NyaaScraper:
    def get_metadata(self, url):
        metadata = {}
        try:
            response = requests.get(url)
            response.raise_for_status()
        except requests.exceptions.RequestException:
            return None
        soup = BeautifulSoup(response.text, "html.parser")

        title = soup.title.string.strip()
        release_name = title.removesuffix(":: Nyaa").strip()

        metadata["release name"] = release_name

        magnet = soup.select_one("div.panel-footer.clearfix a[href^='magnet']")
        magnet = magnet.get("href")
        metadata["magnet"] = magnet

        rows_list = []
        for row in soup.select("div.panel-body .row"):
            for line in row.text.strip().splitlines():
                if line.strip():
                    rows_list.append(line.strip())
        for i, element in enumerate(rows_list):
            match element:
                case "Category:":
                    if i + 1 < len(rows_list):
                        metadata["category"] = rows_list[i + 1]
                case "Date:":
                    if i + 1 < len(rows_list):
                        metadata["date"] = rows_list[i + 1]
                case "Submitter:":
                    if i + 1 < len(rows_list):
                        metadata["submitter"] = rows_list[i + 1]
                case "Seeders:":
                    if i + 1 < len(rows_list):
                        metadata["seeders"] = rows_list[i + 1]
                case "Leechers:":
                    if i + 1 < len(rows_list):
                        metadata["leechers"] = rows_list[i + 1]
                case "File size:":
                    if i + 1 < len(rows_list):
                        metadata["file size"] = rows_list[i + 1]
                case "Completed:":
                    if i + 1 < len(rows_list):
                        metadata["completed"] = rows_list[i + 1]

        files = soup.select_one("div.torrent-file-list.panel-body")
        metadata["files"] = self.format_files(files)

        return metadata

    def format_files(self, files):
        items = []

        if files.name == "div":
            files = files.find("ul")
            if not files:
                return items

        for list_item in files.find_all("li", recursive=False):
            folder = list_item.find("a", class_="folder")
            if folder:
                folder_name = folder.get_text(strip=True)
                nested_ul = list_item.find("ul")
                contents = self.format_files(nested_ul) if nested_ul else []
                items.append(
                    {"type": "folder", "name": folder_name, "contents": contents}
                )
            else:
                file_icon = list_item.find("i", class_="fa-file")
                size_span = list_item.find("span", class_="file-size")
                if file_icon:
                    file_name = (
                        file_icon.next_sibling.strip() if file_icon.next_sibling else ""
                    )
                    file_size = size_span.get_text(strip=True) if size_span else None
                    items.append({"type": "file", "name": file_name, "size": file_size})

        return items


if __name__ == "__main__":
    print(NyaaScraper().get_metadata("https://nyaa.si/view/1960108"))
    # print(NyaaScraper().get_metadata("https://nyaa.si/view/1577473"))

File: src/nyaa_scraper.ts
import browser from "webextension-polyfill";

export interface FileItem {
    type: "file";
    name: string;
    size: string | null;
}

export interface FolderItem {
    type: "folder";
    name: string;
    contents: (FileItem | FolderItem)[];
}

export type NyaaFileEntry = FileItem | FolderItem;

export interface NyaaMetadata {
    releaseName: string;
    magnet: string;
    url?: string;
    category?: string;
    date?: string;
    submitter?: string;
    seeders?: string;
    leechers?: string;
    fileSize?: string;
    completed?: string;
    files: NyaaFileEntry[];
}

interface FetchResponse {
    ok: boolean;
    text: string | null;
}

// Bypasses CORS by delegating fetch to background script
async function backgroundFetch(url: string): Promise<FetchResponse> {
    try {
        const response = await browser.runtime.sendMessage({ type: "fetch", url });
        return response as FetchResponse || { ok: false, text: null };
    } catch (error) {
        console.error("[NyaaScraper] Message error:", error);
        return { ok: false, text: null };
    }
}

const METADATA_FIELDS: Record<string, keyof NyaaMetadata> = {
    "Category:": "category",
    "Date:": "date",
    "Submitter:": "submitter",
    "Seeders:": "seeders",
    "Leechers:": "leechers",
    "File size:": "fileSize",
    "Completed:": "completed",
};

export class NyaaScraper {
    async getMetadata(url: string): Promise<NyaaMetadata | null> {
        const metadata: Partial<NyaaMetadata> = {};

        try {
            const response = await backgroundFetch(url);
            if (!response.ok || !response.text) {
                return null;
            }
            const html = response.text;

            const parser = new DOMParser();
            const soup = parser.parseFromString(html, "text/html");

            const magnetElement = soup.querySelector("div.panel-footer.clearfix a[href^='magnet']");
            if (!magnetElement) {
                return null;
            }
            const magnet = magnetElement.getAttribute("href");
            if (!magnet) {
                return null;
            }

            const titleElement = soup.querySelector("title");
            const title = titleElement?.textContent || "";
            const releaseName = title.replace(":: Nyaa", "").trim();

            metadata.releaseName = releaseName;
            metadata.magnet = magnet;

            const rowsList: string[] = [];
            const rows = soup.querySelectorAll("div.panel-body .row");
            for (const row of rows) {
                const lines = (row.textContent || "").trim().split("\n");
                for (const line of lines) {
                    const trimmed = line.trim();
                    if (trimmed) {
                        rowsList.push(trimmed);
                    }
                }
            }



            for (let i = 0; i < rowsList.length; i++) {
                const element = rowsList[i];
                const key = METADATA_FIELDS[element];
                if (key && i + 1 < rowsList.length) {
                    // Type assertion needed because metadata is Partial<NyaaMetadata> 
                    // and values are strictly strings, but NyaaMetadata has non-string 'files'
                    (metadata as any)[key] = rowsList[i + 1];
                }
            }

            const filesDiv = soup.querySelector("div.torrent-file-list.panel-body");
            metadata.files = filesDiv ? this.formatFiles(filesDiv) : [];

            return metadata as NyaaMetadata;
        } catch {
            return null;
        }
    }

    formatFiles(files: Element): NyaaFileEntry[] {
        const items: NyaaFileEntry[] = [];

        let ul: Element | null = files;
        if (files.tagName === "DIV") {
            ul = files.querySelector("ul");
            if (!ul) {
                return items;
            }
        }

        const listItems = ul.querySelectorAll(":scope > li");
        for (const listItem of listItems) {
            const folder = listItem.querySelector("a.folder");
            if (folder) {
                const folderName = folder.textContent?.trim() || "";
                const nestedUl = listItem.querySelector("ul");
                const contents = nestedUl ? this.formatFiles(nestedUl) : [];
                items.push({
                    type: "folder",
                    name: folderName,
                    contents: contents,
                });
            } else {
                const fileIcon = listItem.querySelector("i.fa-file");
                const sizeSpan = listItem.querySelector("span.file-size");
                if (fileIcon) {
                    const nextSibling = fileIcon.nextSibling;
                    const fileName = nextSibling && nextSibling.nodeType === Node.TEXT_NODE
                        ? nextSibling.textContent?.trim() || ""
                        : "";
                    const fileSize = sizeSpan?.textContent?.trim() || null;
                    items.push({
                        type: "file",
                        name: fileName,
                        size: fileSize,
                    });
                }
            }
        }

        return items;
    }
}
